File: apple/MuCore/Tests/MuCoreTests/SyncFixtureTest.swift
import XCTest
@testable import MuCore

/// 同步引擎契約測試：跑 contract/fixtures/sync_cases/ 全部案例。
/// 驅動 = script.json 步驟 → 真實暫存目錄樹（顯式 mtime）→ 每步一輪 sync()；
/// Swift 引擎輸出必須與 Python 參考實作的 expected.json byte-identical（sync-rules.md §3）。
final class SyncFixtureTest: XCTestCase {

    func testAllSyncFixtureCasesMatchByteForByte() throws {
        let casesDir = try XCTUnwrap(findDir("contract/fixtures/sync_cases"),
                                     "sync_cases not found")
        let assetsDir = try XCTUnwrap(findDir("contract/fixtures/sync_assets"),
                                      "sync_assets not found")
        let names = try XCTUnwrap(
            try FileManager.default.contentsOfDirectory(atPath: casesDir.path)
                .filter { name -> Bool in
                    var isDir: ObjCBool = false
                    let ok = FileManager.default.fileExists(
                        atPath: casesDir.appendingPathComponent(name).path, isDirectory: &isDir)
                    return ok && isDir.boolValue
                }
                .sorted())
        XCTAssertGreaterThanOrEqual(names.count, 6)

        var failures: [String] = []
        for name in names {
            let caseDir = casesDir.appendingPathComponent(name)
            let expected = try String(contentsOf: caseDir.appendingPathComponent("expected.json"),
                                      encoding: .utf8)
            let script = try XCTUnwrap(
                JSONSerialization.jsonObject(
                    with: try Data(contentsOf: caseDir.appendingPathComponent("script.json")))
                as? [String: Any])
            let tmp = FileManager.default.temporaryDirectory
                .appendingPathComponent("mu-sync-\(UUID().uuidString)", isDirectory: true)
            let root = tmp.appendingPathComponent("lib", isDirectory: true)
            try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
            defer { try? FileManager.default.removeItem(at: tmp) }

            let engine = SyncEngine(provider: LocalFolderProvider(root: root))
            var reports: [CanonicalJson.JSONValue] = []
            for step in try XCTUnwrap(script["steps"] as? [[String: Any]]) {
                var deleteAfter: [String] = []
                for op in try XCTUnwrap(step["ops"] as? [[String: Any]]) {
                    try applyOp(root: root, assetsDir: assetsDir, op: op, deleteAfter: &deleteAfter)
                }
                let r = try engine.sync(afterDelta: deleteAfter.isEmpty ? nil : {
                    for p in deleteAfter {
                        try? FileManager.default.removeItem(
                            at: root.appendingPathComponent(p))
                    }
                })
                reports.append(engine.canonical(r))
            }
            let actual = CanonicalJson.render(.array(reports))
            if expected != actual {
                failures.append("case [\(name)]\n--- expected ---\n\(expected)\n" +
                    "--- actual ---\n\(actual)")
            }
        }
        if !failures.isEmpty {
            XCTFail("\(failures.count)/\(names.count) sync cases drifted:\n\n" +
                failures.joined(separator: "\n\n========\n\n"))
        }
    }

    private func applyOp(
        root: URL, assetsDir: URL, op: [String: Any], deleteAfter: inout [String]
    ) throws {
        let fm = FileManager.default
        func s(_ k: String) -> String { op[k] as! String }
        switch s("op") {
        case "write":
            let url = root.appendingPathComponent(s("path"))
            try fm.createDirectory(at: url.deletingLastPathComponent(),
                                   withIntermediateDirectories: true)
            let data: Data
            if let asset = op["asset"] as? String {
                data = try Data(contentsOf: assetsDir.appendingPathComponent(asset))
            } else {
                data = s("text").data(using: .utf8)!
            }
            try data.write(to: url)
            try fm.setAttributes(
                [.modificationDate: Date(timeIntervalSince1970: op["mtime"] as! Double)],
                ofItemAtPath: url.path)
        case "delete":
            try fm.removeItem(at: root.appendingPathComponent(s("path")))
        case "rename":
            let dst = root.appendingPathComponent(s("to"))
            try fm.createDirectory(at: dst.deletingLastPathComponent(),
                                   withIntermediateDirectories: true)
            try fm.moveItem(at: root.appendingPathComponent(s("from")), to: dst)
        case "touch":
            try fm.setAttributes(
                [.modificationDate: Date(timeIntervalSince1970: op["mtime"] as! Double)],
                ofItemAtPath: root.appendingPathComponent(s("path")).path)
        case "delete_after_delta":
            deleteAfter.append(s("path"))
        default:
            XCTFail("unknown op: \(s("op"))")
        }
    }

    private func findDir(_ rel: String) -> URL? {
        var dir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        while true {
            let c = dir.appendingPathComponent(rel)
            if FileManager.default.fileExists(atPath: c.path) { return c }
            if dir.path == "/" { return nil }
            dir = dir.deletingLastPathComponent()
        }
    }
}

File: android/core/src/test/kotlin/crate/core/SyncFixtureTest.kt
package crate.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import java.io.File
import kotlin.io.path.createTempDirectory
import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.test.fail

/**
 * 同步引擎契約測試：跑 contract/fixtures/sync_cases/ 全部案例。
 * 驅動 = script.json 步驟 → 真實暫存目錄樹（顯式 mtime）→ 每步一輪 sync()；
 * Kotlin 引擎輸出必須與 Python 參考實作的 expected.json byte-identical（sync-rules.md §3）。
 */
class SyncFixtureTest {

    @Test
    fun `all sync fixture cases match expected json byte for byte`() {
        val casesDir = findDir("contract/fixtures/sync_cases") ?: fail("sync_cases not found")
        val assetsDir = findDir("contract/fixtures/sync_assets") ?: fail("sync_assets not found")
        val names = casesDir.listFiles()?.filter { it.isDirectory }?.map { it.name }?.sorted()
            ?: fail("cannot list $casesDir")
        assertTrue(names.size >= 6, "expected >=6 sync cases, got ${names.size}")

        val failures = ArrayList<String>()
        for (name in names) {
            val caseDir = File(casesDir, name)
            val expected = File(caseDir, "expected.json").readBytes()
            val script = Json.parseToJsonElement(
                File(caseDir, "script.json").readText()).jsonObject
            val tmp = createTempDirectory("crate-sync-").toFile()
            try {
                val root = File(tmp, "lib").apply { mkdirs() }
                val engine = SyncEngine(LocalFolderProvider(root))
                val reports = ArrayList<Map<String, Any?>>()
                for (step in script["steps"]!!.jsonArray) {
                    val deleteAfter = ArrayList<String>()
                    for (opEl in step.jsonObject["ops"]!!.jsonArray) {
                        applyOp(root, assetsDir, opEl.jsonObject, deleteAfter)
                    }
                    val r = engine.sync(
                        afterDelta = if (deleteAfter.isEmpty()) null else {
                            { deleteAfter.forEach { File(root, it).delete() } }
                        })
                    reports.add(engine.toCanonical(r))
                }
                val actual = CanonicalJson.render(reports).toByteArray(Charsets.UTF_8)
                if (!expected.contentEquals(actual)) {
                    failures.add(
                        "case [$name]\n--- expected ---\n${expected.decodeToString()}\n" +
                            "--- actual ---\n${actual.decodeToString()}"
                    )
                }
            } finally {
                tmp.deleteRecursively()
            }
        }
        if (failures.isNotEmpty()) {
            fail("${failures.size}/${names.size} sync cases drifted:\n\n" +
                failures.joinToString("\n\n========\n\n"))
        }
    }

    private fun applyOp(
        root: File, assetsDir: File,
        op: JsonObject, deleteAfter: MutableList<String>,
    ) {
        fun s(k: String) = op[k]!!.jsonPrimitive.content
        when (val k = s("op")) {
            "write" -> {
                val p = File(root, s("path"))
                p.parentFile?.mkdirs()
                val data = op["asset"]?.jsonPrimitive?.content
                    ?.let { File(assetsDir, it).readBytes() }
                    ?: op["text"]!!.jsonPrimitive.content.toByteArray(Charsets.UTF_8)
                p.writeBytes(data)
                p.setLastModified(op["mtime"]!!.jsonPrimitive.long * 1000)
            }
            "delete" -> File(root, s("path")).delete()
            "rename" -> {
                val dst = File(root, s("to"))
                dst.parentFile?.mkdirs()
                File(root, s("from")).renameTo(dst)
            }
            "touch" -> File(root, s("path"))
                .setLastModified(op["mtime"]!!.jsonPrimitive.long * 1000)
            "delete_after_delta" -> deleteAfter.add(s("path"))
            else -> fail("unknown op: $k")
        }
    }

    private fun findDir(rel: String): File? {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (dir != null) {
            val c = File(dir, rel)
            if (c.isDirectory) return c
            dir = dir.parentFile
        }
        return null
    }
}
